cni/network/invoker_cns.go

```diff
@@ -132,97 +133,44 @@ func (invoker *CNSIPAMInvoker) Add(addConfig IPAMAddConfig) (IPAMAddResult, erro
-				return IPAMAddResult{}, errors.Wrap(errInvalidArgs, "%w: Gateway address "+info.ncGatewayIPAddress+" from response is invalid")
-			}
-
-			if net.ParseIP(info.podIPAddress).To4() != nil { //nolint:gocritic
-				ncgw, err = getOverlayGateway(ncIPNet)
-				if err != nil {
-					return IPAMAddResult{}, err
-				}
-			} else if net.ParseIP(info.podIPAddress).To16() != nil {
-				ncgw = net.ParseIP(overlayGatewayV6IP)
-			} else {
-				return IPAMAddResult{}, errors.Wrap(err, "No podIPAddress is found: %w")
+		//nolint:exhaustive // ignore exhaustive types check
+		switch info.nicType {
+		case cns.DelegatedVMNIC:
+			if !info.skipDefaultRoutes {
+				numInterfacesWithDefaultRoutes++
 			}
-		}
-
-		// construct ipnet for result
-		resultIPnet := net.IPNet{
-			IP:   ip,
-			Mask: ncIPNet.Mask,
-		}
 
-		if net.ParseIP(info.podIPAddress).To4() != nil {
-			addResult.ipv4Result = &cniTypesCurr.Result{
-				IPs: []*cniTypesCurr.IPConfig{
-					{
-						Address: resultIPnet,
-						Gateway: ncgw,
-					},
-				},
-				Routes: []*cniTypes.Route{
-					{
-						Dst: network.Ipv4DefaultRouteDstPrefix,
-						GW:  ncgw,
-					},
-				},
+			if err := configureSecondaryAddResult(&info, &addResult, &response.PodIPInfo[i].PodIPConfig); err != nil {
+				return IPAMAddResult{}, err
 			}
-		} else if net.ParseIP(info.podIPAddress).To16() != nil {
-			addResult.ipv6Result = &cniTypesCurr.Result{
-				IPs: []*cniTypesCurr.IPConfig{
-					{
-						Address: resultIPnet,
-						Gateway: ncgw,
-					},
-				},
-				Routes: []*cniTypes.Route{
-					{
-						Dst: network.Ipv6DefaultRouteDstPrefix,
-						GW:  ncgw,
-					},
-				},
+		default:
+			// only count dualstack interface once
+			if addResult.defaultInterfaceInfo.ipResult == nil && !info.skipDefaultRoutes {
+				numInterfacesWithDefaultRoutes++
 			}
-		}
 
-		// get the name of the primary IP address
-		_, hostIPNet, err := net.ParseCIDR(info.hostSubnet)
-		if err != nil {
-			return IPAMAddResult{}, fmt.Errorf("unable to parse hostSubnet: %w", err)
-		}
-
-		addResult.hostSubnetPrefix = *hostIPNet
-
-		// set subnet prefix for host vm
-		// setHostOptions will execute if IPAM mode is not v4 overlay and not dualStackOverlay mode
-		// TODO: Remove v4overlay and dualstackoverlay options, after 'overlay' rolls out in AKS-RP
-		if (invoker.ipamMode != util.V4Overlay) && (invoker.ipamMode != util.DualStackOverlay) && (invoker.ipamMode != util.Overlay) {
-			if err := setHostOptions(ncIPNet, addConfig.options, &info); err != nil {
+			overlayMode := (invoker.ipamMode == util.V4Overlay) || (invoker.ipamMode == util.DualStackOverlay) || (invoker.ipamMode == util.Overlay)
+			if err := configureDefaultAddResult(&info, &addConfig, &addResult, overlayMode); err != nil {
 				return IPAMAddResult{}, err
 			}
 		}
 	}
 
+	// Make sure default routes exist for 1 interface
+	if numInterfacesWithDefaultRoutes != 1 {
+		return IPAMAddResult{}, errors.New("Add result requires exactly one interface with default routes")
```

Comment:
can we define a new error variable for this error?

Reply:
added new errInvalidDefaultRouting